bundles/org.openhab.binding.plugwiseha/src/main/java/org/openhab/binding/plugwiseha/internal/discovery/PlugwiseHADiscoveryService.java

```diff
@@ -0,0 +1,191 @@
+ * http://www.eclipse.org/legal/epl-2.0
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ */
+package org.openhab.binding.plugwiseha.internal.discovery;
+
+import static org.openhab.binding.plugwiseha.internal.PlugwiseHABindingConstants.*;
+
+import java.util.HashMap;
+import java.util.Map;
+import java.util.concurrent.ScheduledFuture;
+import java.util.concurrent.TimeUnit;
+
+import org.eclipse.jdt.annotation.NonNullByDefault;
+import org.eclipse.jdt.annotation.Nullable;
+import org.openhab.binding.plugwiseha.internal.PlugwiseHABindingConstants;
+import org.openhab.binding.plugwiseha.internal.api.exception.PlugwiseHAException;
+import org.openhab.binding.plugwiseha.internal.api.model.PlugwiseHAController;
+import org.openhab.binding.plugwiseha.internal.api.model.dto.Appliance;
+import org.openhab.binding.plugwiseha.internal.api.model.dto.DomainObjects;
+import org.openhab.binding.plugwiseha.internal.api.model.dto.Location;
+import org.openhab.binding.plugwiseha.internal.handler.PlugwiseHABridgeHandler;
+import org.openhab.core.config.discovery.AbstractDiscoveryService;
+import org.openhab.core.config.discovery.DiscoveryResult;
+import org.openhab.core.config.discovery.DiscoveryResultBuilder;
+import org.openhab.core.thing.ThingUID;
+import org.openhab.core.thing.binding.ThingHandler;
+import org.openhab.core.thing.binding.ThingHandlerService;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * The {@link PlugwiseHADiscoveryService} class is capable of discovering the
+ * available data from the Plugwise Home Automation gateway
+ *
+ * @author Bas van Wetten - Initial contribution
+ * @author Leo Siepel - finish initial contribution
+ */
+@NonNullByDefault
+public class PlugwiseHADiscoveryService extends AbstractDiscoveryService implements ThingHandlerService {
+
+    private final Logger logger = LoggerFactory.getLogger(PlugwiseHADiscoveryService.class);
+    private static final int TIMEOUT_SECONDS = 5;
+    private static final int REFRESH_SECONDS = 600;
+    private @Nullable PlugwiseHABridgeHandler bridgeHandler;
+    private @Nullable ScheduledFuture<?> discoveryFuture;
+
+    public PlugwiseHADiscoveryService() {
+        super(SUPPORTED_THING_TYPES_UIDS, TIMEOUT_SECONDS, true);
+    }
+
+    @Override
+    protected synchronized void startScan() {
+        try {
+            discoverDomainObjects();
+        } catch (PlugwiseHAException e) {
+            // Ignore silently
+        }
+    }
+
+    @Override
+    protected synchronized void stopScan() {
+        super.stopScan();
+        removeOlderResults(getTimestampOfLastScan());
+    }
+
+    @Override
+    protected void startBackgroundDiscovery() {
+        logger.debug("Start Plugwise Home Automation background discovery");
+        if (discoveryFuture == null || discoveryFuture.isCancelled()) {
+            discoveryFuture = scheduler.scheduleWithFixedDelay(this::startScan, 30, REFRESH_SECONDS, TimeUnit.SECONDS);
+        }
```

Comment:
The field can be null at this point, because it isn't inittialized in the constructor. So, you need to check if it is not null before accessing it: Example: if (job != null). But this would result in a compiler warning, as it is a field and could theoretically be modified by another thread exactly between the null check and the access in the next line. To mitigate this, you can store the field to a local variable before checking it:
```
Future<?> localPollingJob = pollingJob;
if (localPollingJob != null) {
    localPollingJob.cancel(true);
}
```

Reply:
I think the example is about cancelling. The comments are placed at scheduling. I think i fixed it, but could you please double check if this is what you meant?